spec/controllers/products_controller_spec.rb

```diff
@@ -0,0 +1,258 @@
+require 'rails_helper'
+require 'support/app_controller'
+
+RSpec.describe ProductsController, type: :controller do
+  describe "GET 'new'" do
+    it "renders new view" do
+      get :new
+      expect(subject).to render_template :new
+    end
+  end
+
+  describe "POST 'create'" do
+    let(:user) do
+      User.create(username: "Mister",
+                  email_address: "stud@manly.com",
+                  password: "pw",
+                  password_confirmation: "pw")
+    end
+
+    let(:good_params) do
+      {
+        product:{
+          name: "necklace",
+          price: 10,
+          user_id: 2,
+          stock: 3
+        },
+        categories: []
+      }
+    end
+
+    let(:bad_params) do
+      {
+        product:{
+          name: nil,
+          price: nil
+        },
+        categories: []
+      }
+    end
+
+    before :each do
+      session[:user_id] = user.id
+    end
+
+    it "redirects to products index page when good params are passed" do
+      post :create, good_params.merge(id: 1)
+      expect(subject).to redirect_to product_path(user.id)
+    end
+
+    it "renders the edit template when bad params are passed" do
+      post :create, bad_params
+      expect(subject).to render_template :new
+    end
+  end
+
+  describe "GET 'edit'" do
+    let(:product) do
+      Product.create(name: "necklace", price: 10, user_id: 2, stock: 3)
+    end
+
+    it "renders edit view" do
+      get :edit, id: product.id
+      expect(subject).to render_template :edit
+    end
+  end
+
+  describe "GET 'show'" do
+    let(:show_product) do
+      Product.create(name: "necklace", price: 10, user_id: 2, stock: 3)
+    end
+
+    it "renders the show view" do
+      get :show, id: show_product.id
+      expect(subject).to render_template :show
+    end
+  end
+
+  describe "PATCH 'update'" do
+    before :each do
+        request.env["HTTP_REFERER"] = "back"
+    end
+
+    let(:current_user) do
+      User.create(username: "FancyPants",
+                  email_address: "fancypants@fancypants.com",
+                  password: "123",
+                  password_confirmation: "123")
+      end
+
+    before :each do
+      session[:user_id] = current_user.id
+    end
```

Comment:
It makes it easier to read if all your `before :each` code is in the same place.

Line 92 could be moved up into the `before :each` block on line 80-82.